retrofit/src/test/java/retrofit2/CallTest.java

```diff
@@ -1125,9 +1125,49 @@
     try {
       call.request();
       fail();
-    } catch (OutOfMemoryError e) {
-      assertThat(e).hasMessage("Broken!");
+    } catch (AssertionError expected) {
+      assertThat(expected).hasMessage("Broken!");
     }
     assertThat(writeCount.get()).isEqualTo(1);
   }
+
+  @Test public void unrecoverableErrorsAreNotCaughtForCallback() throws Exception {
+    Retrofit retrofit = new Retrofit.Builder()
+        .baseUrl(server.url("/"))
+        .addConverterFactory(new ToStringConverterFactory())
+        .build();
+    Service service = retrofit.create(Service.class);
+
+    server.enqueue(new MockResponse());
+
+    final AtomicInteger writeCount = new AtomicInteger();
+    Object a = new Object() {
+      @Override public String toString() {
+        writeCount.incrementAndGet();
+        throw new OutOfMemoryError("Broken!");
+      }
+    };
+    Call<String> call = service.postRequestBody(a);
+
+    try {
+      call.enqueue(new Callback<String>() {
+        @Override public void onResponse(Call<String> call, Response<String> response) {
+        }
+
+        @Override public void onFailure(Call<String> call, Throwable t) {
+        }
```

Comment:
add a `fail()` inside `onFailure()`? Just to verify that error is not delivered to a listener before throw

Reply:
yeah, i'd normally do that, but these callbacks are hit asynchronously, so you'd never be able to test that...

Reply:
i could add a check to make sure the callback isn't hit synchronously, but that might confuse the test...

Reply:
Yep, that was the point, check that error is not synchronously delivered before throw, it's a nit and up2u of course :)

Reply:
yeah, i liked it after seeing it. done, thanks!